libs/datamodel/core/src/common/preview_features.rs

```diff
@@ -1,57 +1,155 @@
-const NAPI: &str = "napi";
-const SELECT_RELATION_COUNT: &str = "selectRelationCount";
-const ORDER_BY_AGGREGATE_GROUP: &str = "orderByAggregateGroup";
-
-// deprecated preview features
-const ATOMIC_NUMBER_OPERATIONS: &str = "atomicNumberOperations";
-const AGGREGATE_API: &str = "aggregateApi";
-const MIDDLEWARES: &str = "middlewares";
-const DISTINCT: &str = "distinct";
-const UNCHECKED_SCALAR_INPUTS: &str = "uncheckedScalarInputs";
-
-pub const DATASOURCE_PREVIEW_FEATURES: &[&str] = &[];
-
-pub const GENERATOR_PREVIEW_FEATURES: &[&str] = &[
-    SQL_SERVER,
-    ORDER_BY_RELATION,
-    NAPI,
-    SELECT_RELATION_COUNT,
-    ORDER_BY_AGGREGATE_GROUP,
-];
-
-/// Hidden preview features are valid features, but are not propagated into the tooling
-/// (as autocomplete or similar) or into error messages (eg. showing a list of valid features).
-pub const HIDDEN_GENERATOR_PREVIEW_FEATURES: &[&str] = &[MONGODB];
-
-pub const DEPRECATED_GENERATOR_PREVIEW_FEATURES: &[&str] = &[
-    ATOMIC_NUMBER_OPERATIONS,
-    AGGREGATE_API,
-    MIDDLEWARES,
-    NATIVE_TYPES,
-    DISTINCT,
-    CONNECT_OR_CREATE,
-    TRANSACTION_API,
-    UNCHECKED_SCALAR_INPUTS,
-    GROUP_BY,
-    CREATE_MANY,
-];
-
-pub const DEPRECATED_DATASOURCE_PREVIEW_FEATURES: &[&str] = &[];
-
-/// Returns all current features available.
-pub fn current_features() -> Vec<&'static str> {
-    let mut features = Vec::from(GENERATOR_PREVIEW_FEATURES);
-    features.extend(HIDDEN_GENERATOR_PREVIEW_FEATURES);
-    features
+use lazy_static::lazy_static;
+use paste::paste;
+use serde::{Serialize, Serializer};
+use PreviewFeature::*;
+
+macro_rules! features {
+    ($( $variant:ident ),*) => {
+        #[derive(Debug, Copy, Clone, PartialEq)]
+        pub enum PreviewFeature {
+            $(
+                $variant,
+            )*
+        }
+
+        impl PreviewFeature {
+            pub fn parse_opt(s: &str) -> Option<Self> {
+                paste! {
+                    let parsed = match s.to_lowercase().as_str() {
+                        $(
+                            stringify!([<$variant:lower>]) => Self::$variant,
+                        )*
+                        _ => return None,
+                    };
+                }
+
+                Some(parsed)
+            }
+        }
+
+        impl ToString for PreviewFeature {
+            fn to_string(&self) -> String {
+                match self {
+                    $(
+                        Self::$variant => decapitalize(stringify!($variant)),
+                    )*
+                }
+            }
+        }
+    };
+}
+
+// (Usually) Append-only list of features.
+features!(
+    ConnectOrCreate,
+    TransactionApi,
+    NativeTypes,
+    GroupBy,
+    CreateMany,
+    AtomicNumberOperations,
+    AggregateApi,
+    Middlewares,
+    Distinct,
+    UncheckedScalarInputs,
+    SqlServer,
+    MongoDb,
+    OrderByRelation,
+    NApi,
+    SelectRelationCount,
+    OrderByAggregateGroup
+);
```

Comment:
Since these are all constants, did you consider using bitflags for this? We have an example [here](https://github.com/prisma/prisma-engines/blob/master/libs/test-setup/src/connectors/tags.rs). It's what I reach for in that sort of circumstances, since it's low boilerplate, typed and easier to deal with in terms of ownership (what you have is `Copy`).

Reply:
I saw that the ME uses bitflags extensively, I find them to be far less readable than simple unary enums.